### PR TITLE
Fix issue #187

### PR DESCRIPTION
**Pull request by elfman**

### CODE REVIEW

File: src/shared/components/challenge-listing/Filters/FiltersPanel/index.jsx [hunk 1]
```diff
@@ -53,6 +53,17 @@ export default function FiltersPanel({
 
   const mapOps = item => ({ label: item, value: item });
 
+  let filterStateTags = null;
+  if (filterState.tags) {
+    filterStateTags = Array.isArray(filterState.tags) ? filterState.tags.join(',') :
+      Object.keys(filterState.tags).map(key => filterState.tags[key]).join(',');
+  }
```
Comment (birdofpreyru) [inline src/shared/components/challenge-listing/Filters/FiltersPanel/index.jsx]: 1. For such checks we prefer to use lodash functions.
2. Is really possible that we can get either array either object here? It looks strange.

File: src/shared/utils/challenge-listing/filter.js [hunk 2]
```diff
@@ -113,7 +113,9 @@ function filterBySubtracks(challenge, state) {
    * it probably does not work in all cases. It should be double-checked,
    * why challenge subtracks in challenge objects are different from those
    * return from the API as the list of possible subtracks. */
-  const filterSubtracks = state.subtracks.map(item =>
+  const subtracks = Array.isArray(state.subtracks) ? state.subtracks :
+    Object.keys(state.subtracks).map(key => state.subtracks[key]);
+  const filterSubtracks = subtracks.map(item =>
```
Comment (birdofpreyru) [inline src/shared/utils/challenge-listing/filter.js]: Especially here. It is just documented in the module's head comment that `tags` and `subtracks` must be arrays. So if it gets called with object, it is an error to be fixed somewhere else.

**Comment by elfman** [comment]
@birdofpreyru I found another way to solve this bug, please check again

File: src/server/server.js [hunk 3]
```diff
@@ -34,6 +35,8 @@ const app = express();
  * fix. */
 global.atob = atob;
 
+app.set('query parser', str => qs.parse(str, { arrayLimit: 1000 }));
```
Comment (birdofpreyru) [inline src/server/server.js]: @elfman It looks like you found the correct reason behind the issue, but I don't like this way solving it. The reason `qs` limits possible array lengths and turns array elements with indices larger than 20 into objects is the following: say you set limit 1000, then it is legit to send query param x[999]=... and the backend will have to create an array with 1000 elements for this param. I guess, it opens a very easy way to spam the server, if somebody wants to: you just send in a query like x[999]=a&y[999]=b&... and server will have lots of fun creating a big array for each of params.
So, I guess, we better just account in the reducer that long arrays are actually parsed as objects.

**Comment by elfman** [comment]
Updated

**Pull request merged**